### PR TITLE
SWIG binding for C++ TemporalMemory

### LINKED ISSUE #2257: SWIG binding for C++ TemporalMemory
Linked to core PR https://github.com/numenta/nupic.core/pull/409
Initial work in https://github.com/rcrowder/nupic/tree/TM-bindings

### PR DESCRIPTION
**Pull request by rcrowder**
Fixes #2257 
(Merged) Relies upon https://github.com/numenta/nupic.core/commit/a23a6df1aff8e475ec6995d8b6f48cae83e8dbb4

### CODE REVIEW

File: nupic/research/monitor_mixin/temporal_memory_monitor_mixin.py [hunk 1]
```diff
@@ -327,7 +331,15 @@ def compute(self, activeColumns, sequenceLabel=None, **kwargs):
...
-    super(TemporalMemoryMonitorMixin, self).compute(activeColumns, **kwargs)
+    # Use the line below for the Python temporal memory (TM)
+    # super(TemporalMemoryMonitorMixin, self).compute(activeColumns, **kwargs)
+
+    # Use the next five lines for the C++ TM
+    array = numpy.array(list(activeColumns), dtype='int32')
+    if kwargs is not None and 'learn' in kwargs:
+      super(TemporalMemoryMonitorMixin, self).compute(len(array), array, kwargs['learn'])
+    else:
+      super(TemporalMemoryMonitorMixin, self).compute(len(array), array, True)
```
Comment (rcrowder) [inline nupic/research/monitor_mixin/temporal_memory_monitor_mixin.py]: This is ugly! Three differing entry points. I wonder if there is a way to combine them?

File: nupic/research/monitor_mixin/temporal_memory_monitor_mixin.py [hunk 2]
```diff
@@ -209,10 +215,8 @@ def mmPrettyPrintConnections(self):
           synapseList = []
 
           for synapse in self.connections.synapsesForSegment(seg):
-            (_, sourceCell, permanence) = self.connections.dataForSynapse(
-              synapse)
-
-            synapseList.append((sourceCell, permanence))
+            synapseData = self.connections.dataForSynapse(synapse)
+            synapseList.append((synapseData.presynapticCell, synapseData.permanence))
```
Comment (rcrowder) [inline nupic/research/monitor_mixin/temporal_memory_monitor_mixin.py]: With the C++ TM, this prefers synapseData.presynapticCell.idx (or it prints bad index during connections pretty printing).

File: nupic/test/abstract_temporal_memory_test.py [hunk 3]
```diff
@@ -24,13 +24,19 @@
 from nupic.data.generators.sequence_machine import SequenceMachine
 from nupic.research.monitor_mixin.temporal_memory_monitor_mixin import (
   TemporalMemoryMonitorMixin)
-from nupic.research.temporal_memory import TemporalMemory
+
+# Uncomment one line below to run tests with C++ or Py TM implementation instead
+from nupic.bindings.algorithms import TemporalMemory
+#from nupic.research.temporal_memory import TemporalMemory
```
Comment (rcrowder) [inline nupic/test/abstract_temporal_memory_test.py]: With a local develop build of nupic, this should be the only place to swap between C++ and Python implementations for the integration tests. But the comment above regarding the compute calls issue shows another change required in the TM mixin code to swap between implementations.

File: nupic/test/abstract_temporal_memory_test.py [hunk 4]
```diff
@@ -56,6 +62,7 @@ def init(self, overrides=None):
     """
     params = self._computeTMParams(overrides)
     self.tm = MonitoredTemporalMemory(**params)
+    #self.tm.printParameters() # C++ only
```
Comment (rcrowder) [inline nupic/test/abstract_temporal_memory_test.py]: Can be a quick way to determine which implementation has been imported.

**Comment by rhyolight** [comment]
@rcrowder This is probably obsolete for nupic now that SWIG code is moved to `nupic.core`, right?

**Comment by rcrowder** [comment]
Moving this issue to the Core library, now that SWIG has moved there.

**Pull request closed without merging**